README.md
# flutter_ritl_alert

青岛地铁APP中使用的AlertController flutter版本

[![Version](https://img.shields.io/badge/version-0.0.2-green.svg)](https://pub.dev/packages/flutter_ritl_alert)
 
## 引用

```yaml
dependencies:
 flutter_ritl_alert: ^0.0.2
```

or

```yaml
dependencies:
 flutter_ritl_alert:
   git: git@github.com:RITL/flutter_ritl_alert.git
```


then run `flutter packages get`

## 无按钮
```dart
RITLAlert(
  context: context,
  title: "我是RITLAlert",
  message: "开启一下GPS定位吧。",
).show();
```
<img width="320" src="https://raw.githubusercontent.com/wiki/RITL/flutter_ritl_alert/none.png"></img>


## 默认方式
```dart
RITLAlert(
  context: context,
  title: "我是RITLAlert",
  message: "开启一下GPS定位吧。",
  buttons: [
    RITLAlertAction(
      title: "知道啦",
      type: RITLAlertActionType.normal,
      onPressed: (alert, action) => print("我点击啦知道啦"),
    ),
    RITLAlertAction(
      title: "取消吧",
      type: RITLAlertActionType.destructive,
      onPressed: (alert, action) => print("我点击啦取消吧"),
    ),
  ],
).show();
```
<img width="320" src="https://raw.githubusercontent.com/wiki/RITL/flutter_ritl_alert/default.png"></img>


## 自定义方式
```dart
RITLAlert(
    context: context,
    title: "我是RITLAlert",
    message: "开启一下GPS定位吧。",
    buttons: [
      RITLAlertAction(
        title: "自定义啦",
        type: RITLAlertActionType.custom,
        style: TextStyle(
          color: Color.fromRGBO(47, 189, 75, 1),
          fontWeight: FontWeight.w500,
        ),
        onPressed: (alert, action) => print("我点击自定义"),
      ),
      RITLAlertAction(
        title: "知道啦",
        type: RITLAlertActionType.normal,
        onPressed: (alert, action) => print("我点击啦知道啦"),
      ),
      RITLAlertAction(
        title: "取消吧",
        type: RITLAlertActionType.destructive,
        onPressed: (alert, action) => print("我点击啦取消吧"),
      ),
    ]).show();
```

<img width="320" src="https://raw.githubusercontent.com/wiki/RITL/flutter_ritl_alert/custom2.png"></img>
<img width="320" src="https://raw.githubusercontent.com/wiki/RITL/flutter_ritl_alert/custom1.png"></img>

## 自定义Style
```dart
RITLAlert(
  context: context,
  title: "我是RITLAlert",
  message: "开启一下GPS定位吧。",
  buttons: [
    RITLAlertAction(
      title: "知道啦",
      type: RITLAlertActionType.normal,
      onPressed: (alert, action) => print("我点击啦知道啦"),
    ),
    RITLAlertAction(
      title: "取消吧",
      type: RITLAlertActionType.destructive,
      onPressed: (alert, action) => print("我点击啦取消吧"),
    ),
  ],
  style: RITLAlertStyle(
    backgroundColor: CupertinoColors.black,
    separateColor: CupertinoColors.white,
    titleStyle: TextStyle(color: CupertinoColors.white),
    messageStyle: TextStyle(color: CupertinoColors.white),
  ),
).show();
```
<img width="320" src="https://raw.githubusercontent.com/wiki/RITL/flutter_ritl_alert/custom_UI.png"></img>


CHANGELOG.md
## [0.0.2] - TODO: Rewirite readme.md

* TODO: rewirite readme.md.


test/flutter_ritl_alert_test.dart
import 'package:flutter_ritl_alert/src/ritl_alert.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_ritl_alert/flutter_ritl_alert.dart';

void main() {
  // prints("matcher");
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
    //
    // final alertController = RITLAlert(context: context);
  });
}


lib/src/ritl_alert.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';

import './ritl_alert_style.dart';
import './ritl_alert_action.dart';

/// [青岛地铁]iOS风格下最简易的alertController样式
class RITLAlert {
  /// 标记的id
  final String? id;
  final RITLAlertStyle style;

  final BuildContext context;

  /// 顶部的标题
  final String title;
  final TextStyle titleStyle;

  /// 中间的描述
  final String? message;
  final TextStyle messageStyle;

  /// 下方的按钮
  final List<RITLAlertAction>? buttons;
  final bool onWillPopActive;
  final bool useRootNavigator;
  // final AlertAnimation? alertAnimation;

  /// ARITLAlertlert constructor
  ///
  /// [context], [title] are required.
  RITLAlert(
      {required this.context,
      this.id,
      required this.title,
      this.style = const RITLAlertStyle(),
      this.titleStyle = const TextStyle(),
      this.message = "",
      this.messageStyle = const TextStyle(),
      this.buttons,
      this.onWillPopActive = false,
      this.useRootNavigator = true});

  //展示方法
  Future<bool?> show() async {
    return await showGeneralDialog(
        context: context,
        pageBuilder: (context, animation, secondAnimation) =>
            _alertContainerView());
  }

  //消失
  Future<void> dismiss() async {
    Navigator.of(context, rootNavigator: useRootNavigator).pop();
  }

  /// alert的容器
  Widget _alertContainerView() {
    //存在并且不为空
    var hasMessage = message != null && !(message?.isEmpty ?? true);

    final Widget _child = ConstrainedBox(
      constraints: style.constraints ??
          BoxConstraints(maxHeight: double.infinity, maxWidth: double.infinity),
      //自适应大小
      child: Align(
        alignment: style.alertAlignment,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Container(
              // height: 300,
              width: 280.0,
              decoration: BoxDecoration(
                color: style.backgroundColor,
                borderRadius: style.borderRadius,
              ),
              child: Column(
                children: [
                  Container(
                    padding: style.alertPadding,
                    child: Column(children: [
                      Text(
                        title,
                        textAlign: style.titleTextAlign,
                        style: style.titleStyle,
                      ),
                      SizedBox(height: 8),
                      !hasMessage
                          ? Container()
                          : Text(
                              message!,
                              textAlign: style.messageTextAlign,
                              style: style.messageStyle,
                            ),
                      SizedBox(height: hasMessage ? 8 : 16),
                    ]),
                  ),
                  //底部的按钮
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: _getButtons(),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );

    return _child;
  }

  /// 获得所有的button
  List<Widget> _getButtons() {
    List<Widget> expandedButtons = [];
    var hasButton = !(buttons == null && (buttons?.isEmpty ?? true));
    //如果按钮不存在，添加默认的取消即可
    if (!hasButton) {
      expandedButtons.add(
        Expanded(
          child: GestureDetector(
            child: Container(
              height: 46,
              decoration: BoxDecoration(
                border: Border(
                  top: BorderSide(width: 1, color: style.separateColor),
                ),
              ),
              child: RITLAlertAction(
                title: "知道啦",
                type: RITLAlertActionType.normal,
              ),
            ),
            onTap: () {
              Navigator.pop(context);
            },
          ),
        ),
      );
    }
    //如果存在自定义的按钮
    else {
      var buttons = this.buttons!;
      expandedButtons.addAll(
        buttons.map(
          (button) => Expanded(
            child: GestureDetector(
              child: Container(
                decoration: BoxDecoration(
                    border: Border(
                        top: BorderSide(
                          width: 1,
                          color: style.separateColor,
                        ),
                        left: buttons.indexOf(button) == 0
                            ? BorderSide.none
                            : BorderSide(
                                width: 1,
                                color: style.separateColor,
                              ))),
                height: 46,
                child: button,
              ),
              onTap: () {
                //执行回调
                Navigator.pop(context);
                if (button.onPressed != null) {
                  button.onPressed!(this, button);
                }
              },
            ),
          ),
        ),
      );
    }
    return expandedButtons;
  }
}


lib/src/ritl_alert_style.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class RITLAlertStyle {
  /// 背景色
  final Color backgroundColor;

  /// 底部按钮的分割颜色
  final Color separateColor;

  /// 总体的圆角
  final BorderRadius? borderRadius;

  /// 标题的风格
  final TextStyle titleStyle;
  final TextAlign titleTextAlign;

  /// 内容的风格
  final TextStyle messageStyle;
  final TextAlign messageTextAlign;

  final BoxConstraints? constraints;

  /// 容器的内部缩进
  final EdgeInsets alertPadding;
  final AlignmentGeometry alertAlignment;

  /// Alert style constructor function
  /// The [backgroundColor] parameter sets the background color.
  /// The [separateColor] parameter sets the button's separate color.
  /// The [borderRadius] parameter sets alert's borderRadius. default is 8
  /// The [titleStyle] parameter sets alert title text style.
  /// The [titleTextAlign] parameter sets alignment of the title.
  /// The [messageStyle] parameter sets alert desc text style.
  /// The [messageTextAlign] parameter sets alignment of the desc.
  /// The [constraints] parameter sets Alert size.
  /// The [alertPadding] parameter sets alert area padding.
  /// The [alertAlignment] parameter sets alert area alignment.
  const RITLAlertStyle({
    this.backgroundColor = CupertinoColors.white,
    this.separateColor = const Color.fromRGBO(247, 247, 247, 1),
    this.borderRadius = const BorderRadius.all(Radius.circular(8)),
    this.titleStyle = const TextStyle(
      fontSize: 16,
      fontWeight: FontWeight.w500,
      color: Color.fromRGBO(77, 76, 77, 1),
    ),
    this.titleTextAlign = TextAlign.center,
    this.messageStyle = const TextStyle(
      fontSize: 14,
      fontWeight: FontWeight.w400,
      color: Color.fromRGBO(88, 89, 89, 1),
    ),
    this.messageTextAlign = TextAlign.center,
    // this.buttonAreaPadding = const EdgeInsets.all(20.0),
    this.constraints,
    // this.buttonsDirection = ButtonsDirection.row,
    // this.alertElevation,
    this.alertPadding = const EdgeInsets.fromLTRB(15, 15, 15, 0),
    //alert的参考点
    this.alertAlignment = Alignment.center,
  });
}


lib/src/ritl_alert_action.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_ritl_alert/flutter_ritl_alert.dart';

/// 按钮的类型
enum RITLAlertActionType { normal, destructive, custom }

extension RITLAlertActionExtension on RITLAlertActionType {
  // textStyle
  TextStyle actionTextStyle() {
    switch (this) {
      case RITLAlertActionType.normal:
        return TextStyle(
          color: Color.fromRGBO(89, 89, 89, 1),
          fontSize: 15,
          fontWeight: FontWeight.w400,
        );
      case RITLAlertActionType.destructive:
        return TextStyle(
          color: CupertinoColors.destructiveRed,
          fontSize: 15,
          fontWeight: FontWeight.w400,
        );
      default:
        return TextStyle(
          color: Color.fromRGBO(89, 89, 89, 1),
          fontSize: 14,
          fontWeight: FontWeight.w400,
        );
    }
  }
}

/// Alert的按钮选项
class RITLAlertAction extends StatelessWidget {
  /// id
  final String id;

  /// 显示的文本
  final String title;

  /// 类型的style
  final TextStyle? style;

  /// 按钮的类型,如果是custom则必须传递自定义的style
  final RITLAlertActionType type;

  /// 点击的回调
  final Function(RITLAlert, RITLAlertAction)? onPressed;

  /// 构建方法
  const RITLAlertAction(
      {required this.title,
      this.id = "",
      this.style,
      this.type = RITLAlertActionType.normal,
      this.onPressed});

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        // GestureDetector()
        Expanded(
          child: Text(
            title,
            style: _textStyle(),
            textAlign: TextAlign.center,
          ),
        )
      ],
    );
  }

  _textStyle() {
    if (type == RITLAlertActionType.custom) {
      return style;
    }
    return type.actionTextStyle();
  }
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_ritl_alert/flutter_ritl_alert.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            CupertinoButton(
              child: Text("RITLAlert(无buttons)"),
              onPressed: () {
                _alertWithNoButtons();
              },
            ),
            CupertinoButton(
              child: Text("RITLAlert(default)"),
              onPressed: () {
                _alertWithDefaultButtons();
              },
            ),
            CupertinoButton(
              child: Text("RITLAlert(Custom)"),
              onPressed: () {
                _alertWithCustom();
              },
            ),
            CupertinoButton(
              child: Text("RITLAlert(UI)"),
              onPressed: () {
                _alertWithUI();
              },
            ),
          ],
        ),
      ),
    );
  }

  // 无按钮的弹出
  _alertWithNoButtons() {
    RITLAlert(
      context: context,
      title: "我是RITLAlert",
      message: "开启一下GPS定位吧。",
    ).show();
  }

  // 默认的按钮弹出
  _alertWithDefaultButtons() {
    RITLAlert(
      context: context,
      title: "我是RITLAlert",
      message: "开启一下GPS定位吧。",
      buttons: [
        RITLAlertAction(
          title: "知道啦",
          type: RITLAlertActionType.normal,
          onPressed: (alert, action) => print("我点击啦知道啦"),
        ),
        RITLAlertAction(
          title: "取消吧",
          type: RITLAlertActionType.destructive,
          onPressed: (alert, action) => print("我点击啦取消吧"),
        ),
      ],
    ).show();
  }

  _alertWithCustom() {
    RITLAlert(
        context: context,
        title: "我是RITLAlert",
        message: "开启一下GPS定位吧。",
        buttons: [
          RITLAlertAction(
            title: "自定义啦",
            type: RITLAlertActionType.custom,
            style: TextStyle(
              color: Color.fromRGBO(47, 189, 75, 1),
              fontWeight: FontWeight.w500,
            ),
            onPressed: (alert, action) => print("我点击自定义"),
          ),
          RITLAlertAction(
            title: "知道啦",
            type: RITLAlertActionType.normal,
            onPressed: (alert, action) => print("我点击啦知道啦"),
          ),
          RITLAlertAction(
            title: "取消吧",
            type: RITLAlertActionType.destructive,
            onPressed: (alert, action) => print("我点击啦取消吧"),
          ),
        ]).show();
  }

  //设置UI
  _alertWithUI() {
    RITLAlert(
      context: context,
      title: "我是RITLAlert",
      message: "开启一下GPS定位吧。",
      buttons: [
        RITLAlertAction(
          title: "知道啦",
          type: RITLAlertActionType.normal,
          onPressed: (alert, action) => print("我点击啦知道啦"),
        ),
        RITLAlertAction(
          title: "取消吧",
          type: RITLAlertActionType.destructive,
          onPressed: (alert, action) => print("我点击啦取消吧"),
        ),
      ],
      style: RITLAlertStyle(
        backgroundColor: CupertinoColors.black,
        separateColor: CupertinoColors.white,
        titleStyle: TextStyle(color: CupertinoColors.white),
        messageStyle: TextStyle(color: CupertinoColors.white),
      ),
    ).show();
  }
}
